### ISSUE TITLE
Segmentation fault (core dumped) with gradient updating.

### ISSUE THREAD

**Issue by cyy5358** [pytorch/pytorch#32265]
## 🐛 Bug
When I updating gradient using
```python
loss_back = ...
loss_back.backward()
scheduler.step()
optimizer.step()
```
A segmentation fault arises randomly. A input causing this crash last time may not causing crash this time.
And after the program exits, gpu memory aren't freed as usual.
I debugged it with gdb, it seems that pytorch is to blame.

## To Reproduce

Actually, I run code from [nvidia/step](https://github.com/NVlabs/STEP) getting this bugs. You may want to follow its installing instruction .

1. git clone https://github.com/NVlabs/STEP.git
2. cd STEP/
3. python setup.py build develop # fp16 wasn't used
4. download all ava2.1 datasets from here [ava](https://github.com/cvdfoundation/ava-dataset/tree/c7f4189d19ab70ef442c10ff4b3c675ee7ae1d67)
5. preparing training folloing here [before training](https://github.com/NVlabs/STEP#training-on-ava-dataset)
6. [train](https://github.com/NVlabs/STEP#training)
7. And after running about a fews min or hours(100 ~ 40k iteartions) , bug will arise.
PS: dataset is hug(about 200G) and it use 4 gpus to run. 

gdb debugging result.
```text
Thread 46 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe5a9de700 (LWP 19345)]
0x00007fffbc46cbfb in std::pair<std::__detail::_Node_iterator<c10::Stream, true, false>, bool> std::_Hashtable<c10::Stream, c10::Stream, std::allocator<c10::Stream>, std::__detail::_Identity, std::equal_to<c10::Stream>, std::hash<c10::Stream>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, true, true> >::_M_emplace<c10::Stream const&>(std::integral_constant<bool, true>, c10::Stream const&) ()
   from /home/chenyuyang/anaconda3/envs/py36/lib/python3.6/site-packages/torch/lib/libtorch.so
(gdb) where
#0  0x00007fffbc46cbfb in std::pair<std::__detail::_Node_iterator<c10::Stream, true, false>, bool> std::_Hashtable<c10::Stream, c10::Stream, std::allocator<c10::Stream>, std::__detail::_Identity, std::equal_to<c10::Stream>, std::hash<c10::Stream>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, true, true> >::_M_emplace<c10::Stream const&>(std::integral_constant<bool, true>, c10::Stream const&) ()
   from /home/chenyuyang/anaconda3/envs/py36/lib/python3.6/site-packages/torch/lib/libtorch.so
#1  0x00007fffbc464380 in torch::autograd::Engine::evaluate_function(torch::autograd::NodeTask&) ()
   from /home/chenyuyang/anaconda3/envs/py36/lib/python3.6/site-packages/torch/lib/libtorch.so
#2  0x00007fffbc466234 in torch::autograd::Engine::thread_main(torch::autograd::GraphTask*) () from /home/chenyuyang/anaconda3/envs/py36/lib/python3.6/site-packages/torch/lib/libtorch.so
#3  0x00007fffe6e5ad1a in torch::autograd::python::PythonEngine::thread_init(int) () from /home/chenyuyang/anaconda3/envs/py36/lib/python3.6/site-packages/torch/lib/libtorch_python.so
#4  0x00007fffe691e19d in std::execute_native_thread_routine (__p=0x7ffba4f7a010)
    at /home/nwani/m3/conda-bld/compilers_linux-64_1560109574129/work/.build/x86_64-conda_cos6-linux-gnu/src/gcc/libstdc++-v3/src/c++11/thread.cc:80
#5  0x00007ffff7bbd6db in start_thread (arg=0x7ffe5a9de700) at pthread_create.c:463
#6  0x00007ffff78e688f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

## Environment
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 5.5.0-12ubuntu1) 5.5.0 20171010
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 430.34
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2019.4                      243    defaults
[conda] mkl-service               2.3.0            py36he904b0f_0    defaults
[conda] mkl_fft                   1.0.15           py36ha843d7b_0    defaults
[conda] mkl_random                1.1.0            py36hd6b4f25_0    defaults
[conda] pytorch                   1.3.1           py3.6_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] torchvision               0.4.2                py36_cu100    pytorch

## Additional context

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by cyy5358** [comment]
I find pytorch1.4 released.
Let me have a try.

**Comment by cyy5358** [comment]
I'm back again. I've tried with PyTorch1.1. 

1.1 is fine. Crash do not happen.

**Comment by albanD** [comment]
Hi,

This is most likely the same as https://github.com/pytorch/pytorch/issues/31906
Can you check with the nightly version if it still crashes?? (1.4 branch does not contain this fix)

**Comment by cyy5358** [comment]
> Hi,
> 
> This is most likely the same as #31906
> Can you check with the nightly version if it still crashes?? (1.4 branch does not contain this fix)

Hi, my code works fine with nightly version torch-1.5.0.dev20200117. Thanks for replying.
